application/src/main/scala/com/azavea/franklin/tile/CogAssetNode.scala

```diff
@@ -21,11 +21,14 @@ case class CogAssetNode(asset: StacItemAsset, bands: Seq[Int]) extends TileUtil
 
     val infoFromCache = sync.get[SerializableGeotiffInfo](tiffKey)
     val tiffInfoIO = infoFromCache match {
-      case Some(info) => IO(info)
+      case Some(info) => IO.pure(info)
       case _ => {
-        val info = GeotiffReader.getGeotiffInfo(asset.href)
-        sync.put(tiffKey)(info)
-        IO(info)
+        for {
+          info <- IO.delay(GeotiffReader.getGeotiffInfo(asset.href))
+        } yield {
+          sync.put(tiffKey)(info)
+          info
+        }
```

Comment:
this is similar to what I was saying but I think does a better job -- keeping the `IO.delay` (which I think is the same as `IO.apply` in the for comprehension makes where the effect happens clear without creating a new thunk for the yielded value. Arguably `sync.put` should also be in the comprehension, but I think the [IO tests](https://github.com/typelevel/cats-effect/blob/master/laws/shared/src/test/scala/cats/effect/IOTests.scala#L95-L99) ensure that there's literally no difference if you do, so :man_shrugging: